File: 713/solution.java
// sliding window

class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        int result = 0;
        while(k <= 1) {
            return result;
        }
        int product = 1, left = 0;
        
        for(int right = 0; right < nums.length; right++) {
            product *= nums[right];
            while(product >= k) {
                product /= nums[left++];
            }
            
            result += right - left + 1;
        }
        
        return result;
    }
}

File: 713/solution.rb
# sliding window

# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def num_subarray_product_less_than_k(nums, k)
    result = 0
    return result if k <= 1
    product, left = 1, 0
    
    nums.size.times do |right|
        product *= nums[right]
        while product >= k
            product /= nums[left]
            left += 1
        end
        
        result += right - left + 1
    end
    
    result
end
